### ISSUE TITLE
lib/connections: Parallel dial isn't great in some circumstances

### ISSUE THREAD

**Issue by calmh** [syncthing/syncthing#5152]
For example when we have two addresses for a device - one which is correct but slow to connect, and one which connects quickly but doesn't work. Typically the "doesn't work" one can be an address that gets rejected after connection by IP ACL, or one that is looped back by mis-announcement (#5151) or NAT router.

It's not clear to me how we fix this easily, maybe we can add punishment delay to addresses that repeatedly fail after  connecting or something.

**Comment by blindchimp** [comment]
Maybe the real measure of success is the amount of data that gets transferred (in some unit of time after you start your connection protocol) rather than whether the connection is successful?

**Comment by AudriusButkevicius** [comment]
Not really, as if everything is in sync, a new healthy connection might have less love than an old one that transfered more data but no longer works.

**Comment by blindchimp** [comment]
Yea, you are right, looking at a total amount wouldn't be too useful. I was thinking more along the lines of just using a metric that got computed within a fixed time of the connect. A bogus connect as described above gets axed pretty quickly, if I understand it. So it wouldn't be able to transfer much data, maybe even none at all. A connection that was otherwise ok, just slow to get going, would at least be able to transfer something, even if it is non-user data (within a few minutes, just to pick a random number). The original issue didn't really mention anything about long lived "dead" connections. 

Does syncthing have any notion of "connection health", or "likelyhood connection will work" at the moment (other than pings)?

**Comment by AudriusButkevicius** [comment]
I think what is described in the original post is sufficient.

You get -1 points on priority (in terms of what is dialed first) on disconnect, which decays over some amount of time (5 x redial interval?).

We might have to re-jiggle existing priorities for this to make more sense, so we don't have 50 point gaps, so that if tcp fails we might pick relay as the better option for an hour or so, I guess we'd retry tcp in some amount of time, and if that is still sucky, hopefully punish that again?

**Comment by AudriusButkevicius** [comment]
Other option would be to send 1mb of data or something like that upon connection to test that it's healthy and usable before we replace the old connection.
